pkg/history/controller_history.go

```diff
@@ -0,0 +1,356 @@
+	DeleteControllerRevision(revision *appsv1.ControllerRevision) error
+	// UpdateControllerRevision updates revision such that its Revision is equal to newRevision. Implementations
+	// may retry on conflict. If the returned error is nil, the update was successful and returned ControllerRevision
+	// is valid. If the returned error is not nil, the update failed and the returned ControllerRevision is invalid.
+	UpdateControllerRevision(revision *appsv1.ControllerRevision, newRevision int64) (*appsv1.ControllerRevision, error)
+	// AdoptControllerRevision attempts to adopt revision by adding a ControllerRef indicating that the parent
+	// Object of parentKind is the owner of revision. If revision is already owned, an error is returned. If the
+	// resource patch fails, an error is returned. If no error is returned, the returned ControllerRevision is
+	// valid.
+	AdoptControllerRevision(parent metav1.Object, parentKind schema.GroupVersionKind, revision *appsv1.ControllerRevision) (*appsv1.ControllerRevision, error)
+	// ReleaseControllerRevision attempts to release parent's ownership of revision by removing parent from the
+	// OwnerReferences of revision. If an error is returned, parent remains the owner of revision. If no error is
+	// returned, the returned ControllerRevision is valid.
+	ReleaseControllerRevision(parent metav1.Object, revision *appsv1.ControllerRevision) (*appsv1.ControllerRevision, error)
+}
+
+// NewHistory returns an instance of Interface that uses client to communicate with the API Server and lister to list
+// ControllerRevisions. This method should be used to create an Interface for all scenarios other than testing.
+func NewHistory(context context.Context, k8sclient client.Client) Interface {
+	return &realHistory{k8sclient, context}
+}
+
+type realHistory struct {
+	client.Client
+	context context.Context
+}
+
+func (rh *realHistory) ListControllerRevisions(parent metav1.Object, selector labels.Selector) ([]*appsv1.ControllerRevision, error) {
+	// List all revisions in the namespace that match the selector
+	revisionList := new(appsv1.ControllerRevisionList)
+	err := rh.List(rh.context, revisionList, client.InNamespace(parent.GetNamespace()), client.MatchingLabelsSelector{Selector: selector})
+	if err != nil {
+		return nil, err
+	}
+	history := revisionList.Items
+	var owned []*appsv1.ControllerRevision
+	for i := range history {
+		ref := metav1.GetControllerOfNoCopy(&history[i])
+		if ref == nil || ref.UID == parent.GetUID() {
+			owned = append(owned, &history[i])
+		}
+
+	}
+	return owned, err
+}
+
+func (rh *realHistory) CreateControllerRevision(parent metav1.Object, revision *appsv1.ControllerRevision) (*appsv1.ControllerRevision, error) {
+	// Clone the input
+	clone := revision.DeepCopy()
+
+	hash := HashControllerRevision(revision)
+	// Update the revisions name
+	clone.Name = ControllerRevisionName(parent.GetName(), hash)
```

Comment:
No need for the clone and setting the name here; setting the name is is already done by `NewControllerRevision`